### PR TITLE
enable `-Werror` by default + warning fixes

### PR DESCRIPTION
**Pull request by OlivierHamel**
Cut from #2009.

Changes:
* `-Werror` is enable by default.

### CODE REVIEW

File: src/ast2ram/provenance/SubproofGenerator.cpp [hunk 1]
```diff
@@ -129,7 +129,8 @@ Own<ram::Operation> SubproofGenerator::addBodyLiteralConstraints(
             auto constraint = mk<ram::Constraint>(
                     BinaryConstraintOp::EQ, std::move(lhs), mk<ram::SubroutineArgument>(i));
             op = mk<ram::Filter>(std::move(constraint), std::move(op));
-        } else if (const auto* adt = as<ast::BranchInit>(arg)) {
+        } else if ([[maybe_unused]] const auto* adt = as<ast::BranchInit>(arg)) {
+            (void)adt;  // GCC 9.2.1 is stupid and ugly and isn't honouring `[[maybe_unused]]`
```
Comment (b-scholz) [inline src/ast2ram/provenance/SubproofGenerator.cpp:133]: Perhaps we rephrase comment to "gcc 9.2.1 does not support [[maybe_unused]]"

File: src/ast2ram/provenance/SubproofGenerator.cpp [hunk 2]
```diff
@@ -231,7 +232,8 @@ Own<ram::Operation> SubproofGenerator::generateReturnInstantiatedValues(const as
         } else if (const auto* rec = as<ast::RecordInit>(arg)) {
             values.push_back(context.translateValue(*valueIndex, rec));
             values.push_back(mk<ram::SubroutineArgument>(i));
-        } else if (const auto* adt = as<ast::BranchInit>(arg)) {
+        } else if ([[maybe_unused]] const auto* adt = as<ast::BranchInit>(arg)) {
+            (void)adt;  // GCC 9.2.1 is stupid and ugly and isn't honouring `[[maybe_unused]]`
```
Comment (b-scholz) [inline src/ast2ram/provenance/SubproofGenerator.cpp:236]: Same here.

**Review by b-scholz — CHANGES_REQUESTED** [review]

File: src/ast2ram/provenance/SubproofGenerator.cpp (hunk 1 above)
Comment (OlivierHamel) [reply src/ast2ram/provenance/SubproofGenerator.cpp:133]: I'll interpret this as 'be tactful'.
Pedantic: The odd thing is that it *does* support `[[maybe_unused]]`, it just doesn't appear to work correctly in the specific case of the vardecl being in an `if` conditional, hence the ire.

**Review by b-scholz — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**